moodler/sections.py

```diff
@@ -0,0 +1,202 @@
+import logging
+import requests
+
+from moodler.consts import REQUEST_FORMAT, validate_response
+from moodler.assignment import get_assignments
+
+logger = logging.getLogger(__name__)
+
+
+# Exceptions for the current module ###
+class SectionsException(Exception):
+    pass
+
+
+class EmptyCoursesList(SectionsException):
+    pass
+
+
+class InvalidCourseName(SectionsException):
+    pass
+
+
+class CoursePrefixNotFound(SectionsException):
+    pass
+
+
+class CourseNotFoundInMoodle(SectionsException):
+    pass
+
+
+class Course(object):
+    def __init__(self, course_id, full_name, short_name):
+        self.id = course_id
+        self.full_name = full_name
+        self.short_name = short_name
+
+    def __repr__(self):
+        return 'Course(courde_id={}, full_name={}, short_name={})'.format(
+            self.id,
+            self.full_name,
+            self.short_name)
+
+
+def core_course_get_courses():
+    """
+    Returns a tuple of ids and course names
+    """
+    response = requests.get(REQUEST_FORMAT.format('core_course_get_courses'))
+
+    validate_response('core_course_get_courses', response.json())
+
+    return response.json()
+
+
+def core_course_get_contents(course_id):
+    """
+    Returns the structure of the course with all resources and topics
+    """
+    response = requests.get(REQUEST_FORMAT.format('core_course_get_contents') +
+                            '&courseid={}'.format(course_id))
+
+    validate_response('core_course_get_contents', response.json())
+
+    return response.json()
+
+
+def list_courses(course_prefix=None):
+    """
+    Create a list of all the courses in the moodle that contain the course
+    prefix received. If the prefix is None, list all courses in the Moodle.
+    :param course_prefix: The prefix that the courses should contain. If not
+    received, list all courses in the Moodle.
+    :return: Returns a list of courses objects.
+    """
+    courses_list = []
+    courses_from_moodle = core_course_get_courses()
+
+    if not courses_from_moodle:
+        raise EmptyCoursesList("The Moodle has returned an empty courses list")
+
+    for course in courses_from_moodle:
+        if course_prefix is not None:
+            if ((course_prefix not in course['shortname']) and
+                    (course_prefix not in course['fullname'])):
+                continue
+
+        courses_list.append(Course(course['id'],
+                                   course['shortname'],
+                                   course['fullname']))
+
+    return courses_list
```

Comment:
Maybe just use the `filter()` function here?

Reply:
The reason I do not use filter is that I need to build the course objects along the way so I have to iterate the list of course anyways...